src/bin/mdbook-xgettext.rs

```diff
@@ -98,6 +104,82 @@ fn create_catalog(ctx: &RenderContext) -> anyhow::Result<Catalog> {
     Ok(catalog)
 }
 
+// Split a main catalog into sub-catalogs, using a provided split
+// depth to separate catalogs based on directory nesting.
+fn split_catalog_on_depth(
+    catalog: &Catalog,
+    split_depth: i64,
+    ctx: &RenderContext,
+    path: &str
+) -> anyhow::Result<HashMap<String, Catalog>> {
+    if split_depth == 0 {
+        return Err(anyhow!(
+            "Catalog doesn't need to be split; depth should be greater than 0."
+        ));
+    };
+
+    // Use the user-provided path to store the messages that don't
+    // belong to any of the separate catalogs. "SUMMARY" is a file
+    // that is used in loading the representation of the Book,
+    // but it does not appear in `ctx.book`, so it has to be
+    // created on its own.
+    let mut catalogs = HashMap::from([
+        (path.to_owned(), Catalog::new(generate_catalog_metadata(&ctx))),
+        ("SUMMARY.md".to_owned(), Catalog::new(generate_catalog_metadata(&ctx))),
+    ]);
+
+    // Grab the path of each chapter, creating a new sub-catalog
+    // if the path meets the depth requirement.
+    for item in ctx.book.iter() {
+        let chapter_path = match item {
+            BookItem::Chapter(chapter) => chapter.path.clone(),
+            BookItem::PartTitle(_) | BookItem::Separator => continue,
+        };
+        if let Some(path) = chapter_path {
+            if path.components().count() <= split_depth as usize {
+                catalogs.insert(
+                    path.display().to_string(),
+                    Catalog::new(generate_catalog_metadata(&ctx))
+                );
+            }
+        }
+    }
+
+    // Source fields in the catalog messages are represented
+    // in a format like `src/{path to file}:{line num}`.
+    // Isolate the path to file and then add the message
+    // to the proper sub-catalog.
+    for message in catalog.messages() {
+        let source = message.source().to_owned();
+
+        let src_path = "src/";
```

Comment:
I don't think we can assume that the path starts with `src/` — it's [configurable](https://rust-lang.github.io/mdBook/format/configuration/general.html#general-metadata).

I think you can get around with this by letting the `RenderContext` drive the process: instead of splitting a catalog after creating it, you create multiple catalogs, one per sub-tree of the book.

Reply:
Fixed!